File: ekodb-client-kt/src/main/kotlin/io/ekodb/client/SchemaBuilder.kt
package io.ekodb.client

import kotlinx.serialization.json.*

/**
 * Fluent builder for field type schemas with constraints and indexes.
 *
 * Example:
 * ```kotlin
 * val field = FieldTypeSchemaBuilder("string")
 *     .required()
 *     .unique()
 *     .textIndex()
 *     .build()
 * ```
 */
class FieldTypeSchemaBuilder(private val fieldType: String) {
    private var required = false
    private var unique = false
    private var default: JsonElement? = null
    private var enums: List<JsonElement>? = null
    private var min: JsonElement? = null
    private var max: JsonElement? = null
    private var regex: String? = null
    private var index: JsonObject? = null

    /** Mark the field as required. */
    fun required() = apply { this.required = true }

    /** Enforce uniqueness on this field. */
    fun unique() = apply { this.unique = true }

    /** Set a default value for the field. */
    fun defaultValue(value: JsonElement) = apply { this.default = value }

    /** Set a default string value. */
    fun defaultValue(value: String) = apply { this.default = JsonPrimitive(value) }

    /** Set a default numeric value. */
    fun defaultValue(value: Number) = apply { this.default = JsonPrimitive(value) }

    /** Set a default boolean value. */
    fun defaultValue(value: Boolean) = apply { this.default = JsonPrimitive(value) }

    /** Restrict field to a set of allowed values. */
    fun enums(values: List<JsonElement>) = apply { this.enums = values }

    /** Restrict field to a set of allowed string values. */
    fun enumStrings(vararg values: String) = apply {
        this.enums = values.map { JsonPrimitive(it) }
    }

    /** Set min/max value constraints. */
    fun range(min: Number? = null, max: Number? = null) = apply {
        min?.let { this.min = JsonPrimitive(it) }
        max?.let { this.max = JsonPrimitive(it) }
    }

    /** Set a regex validation pattern. */
    fun pattern(regex: String) = apply { this.regex = regex }

    /** Add a full-text search index. */
    fun textIndex(language: String = "english", analyzer: String? = null) = apply {
        this.index = buildJsonObject {
            put("type", "text")
            put("language", language)
            analyzer?.let { put("analyzer", it) }
        }
    }

    /**
     * Add a vector similarity search index.
     *
     * @param algorithm Index algorithm ("flat", "hnsw", "ivf")
     * @param metric Distance metric ("cosine", "euclidean", "dotproduct")
     * @param m HNSW parameter for max connections per node
     * @param efConstruction HNSW parameter for construction-time search width
     */
    fun vectorIndex(
        algorithm: String = "flat",
        metric: String = "cosine",
        m: Int = 16,
        efConstruction: Int = 200
    ) = apply {
        this.index = buildJsonObject {
            put("type", "vector")
            put("algorithm", algorithm)
            put("metric", metric)
            put("m", m)
            put("ef_construction", efConstruction)
        }
    }

    /** Add a B-tree index for range queries and exact matches. */
    fun btreeIndex() = apply {
        this.index = buildJsonObject { put("type", "btree") }
    }

    /** Add a hash index for exact matches (faster than B-tree). */
    fun hashIndex() = apply {
        this.index = buildJsonObject { put("type", "hash") }
    }

    /** Build the field type schema as a JsonObject. */
    fun build(): JsonObject = buildJsonObject {
        put("field_type", fieldType)
        put("required", required)
        put("unique", unique)
        default?.let { put("default", it) }
        enums?.let { put("enums", JsonArray(it)) }
        min?.let { put("min", it) }
        max?.let { put("max", it) }
        regex?.let { put("regex", it) }
        index?.let { put("index", it) }
    }
}

/**
 * Fluent builder for collection schemas.
 *
 * Example:
 * ```kotlin
 * val schema = SchemaBuilder()
 *     .addField("name", FieldTypeSchemaBuilder("string").required().build())
 *     .addField("email", FieldTypeSchemaBuilder("string").unique().build())
 *     .build()
 *
 * client.createCollection("users", schema)
 * ```
 */
class SchemaBuilder {
    private val fields = mutableMapOf<String, JsonObject>()
    private var bypassRipple = true
    private var version = 1

    /** Add a field definition from a pre-built JsonObject. */
    fun addField(name: String, fieldSchema: JsonObject) = apply {
        fields[name] = fieldSchema
    }

    /** Add a field definition from a builder (auto-built). */
    fun addField(name: String, builder: FieldTypeSchemaBuilder) = apply {
        fields[name] = builder.build()
    }

    /** Set whether to bypass ripple on collection creation. */
    fun bypassRipple(bypass: Boolean = true) = apply { this.bypassRipple = bypass }

    /** Set the schema version. */
    fun version(version: Int) = apply { this.version = version }

    /** Build the final schema as a JsonObject. */
    fun build(): JsonObject = buildJsonObject {
        put("fields", buildJsonObject {
            fields.forEach { (name, schema) -> put(name, schema) }
        })
        put("version", version)
        put("bypass_ripple", bypassRipple)
    }
}

File: ekodb-client-py/python/ekodb_client/schema_builder.py
"""
SchemaBuilder for defining ekoDB collection schemas with a fluent API.

Example:
    ```python
    from ekodb_client import SchemaBuilder, FieldTypeSchemaBuilder

    schema = (
        SchemaBuilder()
        .add_field("name", FieldTypeSchemaBuilder("string").required().build())
        .add_field("email", FieldTypeSchemaBuilder("string").required().unique().build())
        .add_field("age", FieldTypeSchemaBuilder("number").range(0, 150).build())
        .add_field("bio", FieldTypeSchemaBuilder("string").text_index().build())
        .add_field("embedding", FieldTypeSchemaBuilder("vector").vector_index().build())
        .build()
    )

    await client.create_collection("users", schema)
    ```
"""

from typing import Any, Dict, List, Optional, Union


class FieldTypeSchemaBuilder:
    """Fluent builder for field type schemas with constraints and indexes."""

    def __init__(self, field_type: str):
        """Create a field type schema builder.

        Args:
            field_type: The field type (e.g., "string", "number", "boolean",
                        "vector", "datetime", "uuid", "decimal", "array",
                        "set", "object", "binary", "bytes", "duration")
        """
        self._field_type = field_type
        self._required = False
        self._unique = False
        self._default: Optional[Any] = None
        self._enums: List[Any] = []
        self._min: Optional[Any] = None
        self._max: Optional[Any] = None
        self._regex: Optional[str] = None
        self._index: Optional[Dict[str, Any]] = None

    def required(self) -> "FieldTypeSchemaBuilder":
        """Mark the field as required."""
        self._required = True
        return self

    def unique(self) -> "FieldTypeSchemaBuilder":
        """Enforce uniqueness on this field."""
        self._unique = True
        return self

    def default_value(self, value: Any) -> "FieldTypeSchemaBuilder":
        """Set a default value for the field."""
        self._default = value
        return self

    def enums(self, values: List[Any]) -> "FieldTypeSchemaBuilder":
        """Restrict field to a set of allowed values."""
        self._enums = values
        return self

    def range(
        self, min_val: Optional[Any] = None, max_val: Optional[Any] = None
    ) -> "FieldTypeSchemaBuilder":
        """Set min/max value constraints."""
        self._min = min_val
        self._max = max_val
        return self

    def pattern(self, regex: str) -> "FieldTypeSchemaBuilder":
        """Set a regex validation pattern."""
        self._regex = regex
        return self

    def text_index(
        self, language: str = "english", analyzer: Optional[str] = None
    ) -> "FieldTypeSchemaBuilder":
        """Add a full-text search index."""
        index: Dict[str, Any] = {"type": "text", "language": language}
        if analyzer is not None:
            index["analyzer"] = analyzer
        self._index = index
        return self

    def vector_index(
        self,
        algorithm: str = "flat",
        metric: str = "cosine",
        m: int = 16,
        ef_construction: int = 200,
    ) -> "FieldTypeSchemaBuilder":
        """Add a vector similarity search index.

        Args:
            algorithm: Index algorithm ("flat", "hnsw", "ivf")
            metric: Distance metric ("cosine", "euclidean", "dotproduct")
            m: HNSW parameter for max connections per node
            ef_construction: HNSW parameter for construction-time search width
        """
        self._index = {
            "type": "vector",
            "algorithm": algorithm,
            "metric": metric,
            "m": m,
            "ef_construction": ef_construction,
        }
        return self

    def btree_index(self) -> "FieldTypeSchemaBuilder":
        """Add a B-tree index for range queries and exact matches."""
        self._index = {"type": "btree"}
        return self

    def hash_index(self) -> "FieldTypeSchemaBuilder":
        """Add a hash index for exact matches (faster than B-tree)."""
        self._index = {"type": "hash"}
        return self

    def build(self) -> Dict[str, Any]:
        """Build the field type schema dict."""
        schema: Dict[str, Any] = {
            "field_type": self._field_type,
            "required": self._required,
            "unique": self._unique,
        }

        if self._default is not None:
            schema["default"] = self._default
        if self._enums:
            schema["enums"] = self._enums
        if self._min is not None:
            schema["min"] = self._min
        if self._max is not None:
            schema["max"] = self._max
        if self._regex is not None:
            schema["regex"] = self._regex
        if self._index is not None:
            schema["index"] = self._index

        return schema


class SchemaBuilder:
    """Fluent builder for collection schemas."""

    def __init__(self):
        self._fields: Dict[str, Dict[str, Any]] = {}
        self._bypass_ripple: bool = True
        self._version: int = 1

    def add_field(
        self, name: str, field_schema: Union[Dict[str, Any], "FieldTypeSchemaBuilder"]
    ) -> "SchemaBuilder":
        """Add a field definition to the schema.

        Args:
            name: The field name
            field_schema: Either a dict from FieldTypeSchemaBuilder.build()
                          or a FieldTypeSchemaBuilder instance (auto-built)
        """
        if isinstance(field_schema, FieldTypeSchemaBuilder):
            self._fields[name] = field_schema.build()
        else:
            self._fields[name] = field_schema
        return self

    def bypass_ripple(self, bypass: bool = True) -> "SchemaBuilder":
        """Set whether to bypass ripple on collection creation."""
        self._bypass_ripple = bypass
        return self

    def version(self, version: int) -> "SchemaBuilder":
        """Set the schema version."""
        self._version = version
        return self

    def build(self) -> Dict[str, Any]:
        """Build the final schema dict."""
        return {
            "fields": self._fields,
            "version": self._version,
            "bypass_ripple": self._bypass_ripple,
        }
